File: stack/stack.cxx
#include <stack.h>
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

// ----------------------------------------------------------------------------      
Stack * create_stack(int size)
{
    int * array = (int *) malloc(sizeof(int) * size); 
    assert(array != NULL);

    Stack *stack = new Stack;
    stack->array = array;
    stack->top = -1;
    stack->max_size = size;
    
    return stack;
}
     
// ---------------------------------------------------------------------------- 
void destroy_stack(Stack *stack)
{
    stack->top = NULL;
    stack->max_size = NULL;
    free(stack->array);    
}

// ----------------------------------------------------------------------------
void push(Stack *stack, int element)
{
    // if the stack is full, then grow it to the double 
    if (stack->top == (stack->max_size - 1))
    { 
        stack->max_size *= 2; 
        stack->array = (int*)realloc(stack->array, sizeof(int) * stack->max_size);
        stack->top++;
        stack->array[++stack->top] = element;  

    } 
    else 
    {        
        stack->top++;
        stack->array[stack->top] = element;    
    }
}

// ----------------------------------------------------------------------------
int pop(Stack *stack)
{
    // if the stack is not empty, then return the value at the top
    if (stack->top >= 0)
    {
        int data = stack->array[stack->top];
        stack->top--;
        return data;
    }
    else {
        printf("Error the stack is empty \n");
        return NULL;
    }
    
}

// ----------------------------------------------------------------------------
void print_stack(Stack *stack)
{
    for (int i = 0; i < stack->max_size; i++)
        {
            printf("%d ", *(stack->array + i));
        }    
        
    printf("\n");    
}

File: Prac14/Stack.py
class Stack(object):

    def __init__(self, size):
        self.max_size = size
        self.array = [None] * self.max_size
        self.top = -1        

    def __str__(self):
        """
        print the stack    
        """
        return str([x for x in self.array])

    def push(self, data):
        """
        insert one element to the top of the stack
        """
        if self.top != self.max_size - 1:
            self.top += 1
            self.array[self.top] = data
        else:
            print("Error -> The stack is full")
    
    def pop(self):
        """
        return and remove the data contained at the top of the stack
        """
        if self.top >= 0:
            data =  self.array[self.top]
            self.top -= 1
            return data    
        else:
            print("Error -> The stack is empty")    
        
        
## test the stack        
if __name__ == '__main__':
    stack = Stack(5)
    print("stack =>", stack)
    stack.push(4)
    stack.push(12)
    stack.push(34)
    print("stack =>", stack)
    
    data = stack.pop()
    print("pop = ", data)
    data = stack.pop()
    print("pop = ", data)
    
    print("stack =>", stack)
    stack.push(7)
    stack.push(56)
    stack.push(78)    
    stack.push(98)
    stack.push(1029)
    print("stack =>", stack)
    